geometry/multipolymesh.hpp

```diff
@@ -0,0 +1,286 @@
+     * FacePlaneCrs
+     *
+     * @param[in] faceIdx face index
+     * @returns the coordinate system object
+     */
+    FacePlaneCrs getFaceCrs(std::size_t faceIdx) const;
+
+    /**
+     * Creates coordinate systems for all faces and stores them to `faceCrs`.
+     *
+     * `faceCrs` is cleared in the process
+     */
+    void initalizeFaceCrs();
+
+    /**
+     * Triangulate polygonal faces
+     *
+     * If `faceCrs` is empty, calls getFaceCrs()
+     *
+     * @param[out] triFaceLabels optional labels of faces in triangular mesh
+     *                           (generated if `faceLabels` is not empty)
+     * @returns triangular mesh
+     */
+    Mesh triangulateFaces(std::vector<FaceLabels>* triFaceLabels
+                          = nullptr) const;
+};
+
+// Implementation following
+
+namespace detail
+{
+constexpr double eps = 1e-9;
+}
+
+template <typename FaceLabels>
+FacePlaneCrs MultiPolyMesh<FaceLabels>::getFaceCrs(std::size_t faceIdx) const
+{
+    auto& ob = faces[faceIdx][0];
+    math::Point3 p1 = vertices[ob[0]];
+
+    std::size_t i = 1;
+    math::Point3 p2 = vertices[ob[i]];
+
+    while (math::length(p2 - p1) < detail::eps)
+    {
+        ++i;
+        if (i >= ob.size())
+        {
+            LOGTHROW(err4, std::runtime_error)
+                << "Cannot find another well-distant point";
+        }
+        p2 = vertices[ob[i]];
+    }
+
+    math::Line3 l(p1, p2 - p1);
+
+    math::Point3 p3 = vertices[ob[i]];
+
+    while (math::pointLineDistance(p3, l) < detail::eps)
+    {
+        ++i;
+
+        if (i >= ob.size())
+        {
+            LOGTHROW(err4, std::runtime_error)
+                << "Cannot find another non-collinear point";
+        }
+
+        p3 = vertices[ob[i]];
+    }
+
+    FacePlaneCrs crs(p1, p2, p3);
+
+    // check that the outer boundary has positive area when transformed
+    math::Points2 pts2D;
+    for (auto& v : ob)
+    {
+        pts2D.push_back(crs.to2D(vertices[v]));
+    }
+    if (area(pts2D) > 0) { return crs; }
+    return FacePlaneCrs(p1, p3, p2);
+}
+
+template <typename FaceLabels>
+void MultiPolyMesh<FaceLabels>::initalizeFaceCrs()
+{
+    faceCrs.clear();
+    faceCrs.reserve(faces.size());
+
+    for (std::size_t fIdx = 0; fIdx < faces.size(); ++fIdx)
+    {
+        faceCrs.push_back(getFaceCrs(fIdx));
+    }
+}
+
+/// Helper - convert iterator pair in multipolygon to index of vertex in
+/// MultiPolyMesh
+inline std::size_t itPair2Vertex(const ItPair& itPair,
+                                 const math::MultiPolygon& poly2D,
+                                 const MultiPolyFace& polyFace)
+{
+    std::size_t polyIdx = std::distance(poly2D.begin(), itPair.first);
+    std::size_t ptIdx = std::distance(poly2D[polyIdx].begin(), itPair.second);
+    return polyFace[polyIdx][ptIdx];
+}
+
+template <typename FaceLabels>
+Mesh MultiPolyMesh<FaceLabels>::triangulateFaces(
+    std::vector<FaceLabels>* triFaceLabels) const
```

Comment:
`optional`?

Reply:
Just discussed this with @vaclavblazek and he approved the solution with pointers, but he suggested making the pointer const.